src/mapping/mod.rs

```diff
@@ -0,0 +1,355 @@
+use crate::event::{Event, Value};
+
+pub mod parser;
+pub mod query;
+
+use string_cache::DefaultAtom as Atom;
+
+pub type Result<T> = std::result::Result<T, String>;
+
+pub(self) trait Function: Send + core::fmt::Debug {
+    fn apply(&self, target: &mut Event) -> Result<()>;
+}
+
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub(self) struct Assignment {
+    path: String,
+    function: Box<dyn query::Function>,
+}
+
+impl Assignment {
+    pub(self) fn new(path: String, function: Box<dyn query::Function>) -> Self {
+        Self { path, function }
+    }
+}
+
+impl Function for Assignment {
+    fn apply(&self, target: &mut Event) -> Result<()> {
+        let v = self.function.execute(&target)?;
+        target.as_mut_log().insert(&self.path, v);
+        Ok(())
+    }
+}
+
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub(self) struct Deletion {
+    // TODO: Switch to String once Event API is cleaned up.
+    path: Atom,
+}
+
+impl Deletion {
+    pub(self) fn new(path: String) -> Self {
+        Self {
+            path: Atom::from(path),
+        }
+    }
+}
+
+impl Function for Deletion {
+    fn apply(&self, target: &mut Event) -> Result<()> {
+        target.as_mut_log().remove(&self.path);
+        Ok(())
+    }
+}
+
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub(self) struct OnlyFields {
```

Comment:
Was there discussion of a `ExceptFields` or the reverse of this?

Reply:
Conceptually the `del` function is meant to be the opposite of `only_fields`, but it only supported one path so I've added the ability to specify multiple paths so they behave the same.